Userland/Libraries/LibUnicode/IDNA.cpp

```diff
@@ -0,0 +1,257 @@
+{
+    bool has_error = false;
+    StringBuilder mapped;
+    // 1. Map. For each code point in the domain_name string, look up the status value in Section 5, IDNA Mapping Table, and take the following actions:
+    for (u32 code_point : domain_name) {
+        Optional<Mapping> mapping = get_idna_mapping(code_point);
+        if (!mapping.has_value()) {
+            has_error = true;
+            continue;
+        }
+        switch (translate_status(mapping->status, use_std3_ascii_rules)) {
+        // disallowed: Leave the code point unchanged in the string, and record that there was an error.
+        case MappingStatus::Disallowed:
+            TRY(mapped.try_append_code_point(code_point));
+            has_error = true;
+            break;
+        // ignored: Remove the code point from the string. This is equivalent to mapping the code point to an empty string.
+        case MappingStatus::Ignored:
+            break;
+        // mapped: Replace the code point in the string by the value for the mapping in Section 5, IDNA Mapping Table.
+        case MappingStatus::Mapped:
+            TRY(mapped.try_append(mapping->mapped_to));
+            break;
+        // deviation:
+        case MappingStatus::Deviation:
+            if (transitional_processing == TransitionalProcessing::Yes) {
+                // If Transitional_Processing, replace the code point in the string by the value for the mapping in Section 5, IDNA Mapping Table .
+                TRY(mapped.try_append(mapping->mapped_to));
+            } else {
+                TRY(mapped.try_append_code_point(code_point));
+            }
+            break;
+        // valid: Leave the code point unchanged in the string.
+        case MappingStatus::Valid:
+            TRY(mapped.try_append_code_point(code_point));
+            break;
+
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    // 2. Normalize. Normalize the domain_name string to Unicode Normalization Form C.
+    auto normalized = normalize(mapped.string_view(), NormalizationForm::NFC);
+
+    // 3. Break. Break the string into labels at U+002E ( . ) FULL STOP.
+    auto labels = TRY(normalized.split('.', SplitBehavior::KeepEmpty));
+
+    // 4. Convert/Validate. For each label in the domain_name string:
+    for (auto& label : labels) {
+        // If the label starts with “xn--”:
+        if (label.starts_with_bytes("xn--"sv)) {
+            // 1. Attempt to convert the rest of the label to Unicode according to Punycode [RFC3492]. If that conversion fails, record that there was an error, and continue with the next label.
+            //    Otherwise replace the original label in the string by the results of the conversion.
+            auto punycode = Punycode::decode(label.bytes_as_string_view().substring_view(4));
+            if (punycode.is_error())
+                has_error = true;
```

Comment:
Missing `continue`? "record that there was an error, and continue with the next label"